### PR TITLE
port barf-if-buffer-read-only

### LINKED ISSUE #437: Port barf-if-buffer-read-only to Rust
```c
DEFUN ("barf-if-buffer-read-only", Fbarf_if_buffer_read_only,
				   Sbarf_if_buffer_read_only, 0, 1, 0,
       doc: /* Signal a `buffer-read-only' error if the current buffer is read-only.
If the text under POSITION (which defaults to point) has the
`inhibit-read-only' text property set, the error will not be raised.  */)
  (Lisp_Object position)
{
  if (NILP (position))
    XSETFASTINT (position, PT);
  else
    CHECK_NUMBER (position);

  if (!NILP (BVAR (current_buffer, read_only))
      && NILP (Vinhibit_read_only)
      && NILP (Fget_text_property (position, Qinhibit_read_only, Qnil)))
    xsignal1 (Qbuffer_read_only, Fcurrent_buffer ());
  return Qnil;
}
```

### PR DESCRIPTION
**Pull request by brotzeit**
closes #437

### CODE REVIEW

File: rust_src/src/buffers.rs [hunk 1]
```diff
@@ -118,6 +120,11 @@ impl LispBufferRef {
         LispObject::from(self.name)
     }
 
+    #[inline]
+    pub fn is_read_only(&self) -> LispObject {
```
Comment (shaleh) [inline rust_src/src/buffers.rs]: Is it worth simplifying the Rust code to have this return a bool?

If nil false else true?

File: src/buffer.c [hunk 2]
```diff
@@ -2045,25 +2045,6 @@ set_buffer_if_live (Lisp_Object buffer)
...

-DEFUN ("barf-if-buffer-read-only", Fbarf_if_buffer_read_only,
-				   Sbarf_if_buffer_read_only, 0, 1, 0,
-       doc: /* Signal a `buffer-read-only' error if the current buffer is read-only.
-If the text under POSITION (which defaults to point) has the
-`inhibit-read-only' text property set, the error will not be raised.  */)
-  (Lisp_Object position)
-{
-  if (NILP (position))
-    XSETFASTINT (position, PT);
```
Comment (shaleh) [inline src/buffer.c]: I missed that this is setting position not just returning it.

File: rust_src/src/buffers.rs [hunk 3]
```diff
@@ -4,9 +4,11 @@ use libc::{c_int, c_uchar, c_void, ptrdiff_t};
 use std::{mem, ptr};
 
 use remacs_macros::lisp_fn;
-use remacs_sys::{EmacsInt, Lisp_Buffer, Lisp_Object, Lisp_Overlay, Lisp_Type, Vbuffer_alist};
-use remacs_sys::{make_lisp_ptr, nsberror, set_buffer_internal};
+use remacs_sys::{globals, EmacsInt, Lisp_Buffer, Lisp_Object, Lisp_Overlay, Lisp_Type,
+                 Qbuffer_read_only, Qinhibit_read_only, Qnil, Vbuffer_alist};
```
Comment (shaleh) [inline rust_src/src/buffers.rs]: We have been putting the Qfoo and Vfoo items in their own import statements to help minimize conflicts.

File: src/buffer.c (hunk 2 above)
Comment (brotzeit) [reply src/buffer.c]: Hmmm. I don't see how this is setting point. It looks like this is setting the argument position to point ? As I see it, the function's purpose is only to signal an error if buffer/position is read only.

Comment (shaleh) [reply src/buffer.c]: It isn't. never mind. Your code mimics the intent just fine.

File: rust_src/src/buffers.rs [hunk 4]
```diff
@@ -360,4 +367,26 @@ pub fn set_buffer(buffer_or_name: LispObject) -> LispObject {
     buffer
 }
 
+/// Signal a `buffer-read-only' error if the current buffer is read-only.
+/// If the text under POSITION (which defaults to point) has the
+/// `inhibit-read-only' text property set, the error will not be raised.
+#[lisp_fn(min = "0")]
+pub fn barf_if_buffer_read_only(position: LispObject) -> LispObject {
+    let position = if position.is_nil() { point() } else { position };
+    position.as_fixnum_or_error();
```
Comment (shaleh) [inline rust_src/src/buffers.rs]: XSETFASTINT is natnum not fixnum right?

File: rust_src/src/buffers.rs (hunk 1 above)
Comment (shaleh) [reply rust_src/src/buffers.rs]: @brotzeit ? Is there are reason to have this return a LispObject?

**Comment by shaleh** [comment]
So this was reverted. Here is working code:

    diff --git a/rust_src/remacs-sys/lib.rs b/rust_src/remacs-sys/lib.rs
    index ca40010..f5c8e7f 100755
    --- a/rust_src/remacs-sys/lib.rs
    +++ b/rust_src/remacs-sys/lib.rs
    @@ -1280,6 +1280,11 @@ extern "C" {
         pub fn window_menu_bar_p(w: *const Lisp_Window) -> bool;
         pub fn window_tool_bar_p(w: *const Lisp_Window) -> bool;
     
    +    pub fn Fget_text_property(
    +        position: Lisp_Object,
    +        prop: Lisp_Object,
    +        object: Lisp_Object,
    +    ) -> Lisp_Object;
     }
     
     /// Contains C definitions from the font.h header.
    diff --git a/rust_src/src/buffers.rs b/rust_src/src/buffers.rs
    index 12c6665..15e6923 100644
    --- a/rust_src/src/buffers.rs
    +++ b/rust_src/src/buffers.rs
    @@ -5,8 +5,10 @@ use std::{mem, ptr};
     
     use remacs_macros::lisp_fn;
     use remacs_sys::{EmacsInt, Lisp_Buffer, Lisp_Object, Lisp_Overlay, Lisp_Type, Vbuffer_alist};
    -use remacs_sys::{make_lisp_ptr, nsberror, set_buffer_internal};
    +use remacs_sys::{globals, make_lisp_ptr, nsberror, set_buffer_internal, Fget_text_property};
    +use remacs_sys::{Qbuffer_read_only, Qinhibit_read_only, Qintegerp, Qnil};
     
    +use editfns::point;
     use lisp::{ExternalPtr, LispObject};
     use lisp::defsubr;
     use lists::{car, cdr};
    @@ -22,6 +24,10 @@ pub type LispBufferRef = ExternalPtr<Lisp_Buffer>;
     pub type LispOverlayRef = ExternalPtr<Lisp_Overlay>;
     
     impl LispBufferRef {
    +    pub fn is_read_only(&self) -> bool {
    +        LispObject::from_raw(self.read_only).is_not_nil()
    +    }
    +
         #[inline]
         pub fn zv(self) -> ptrdiff_t {
             self.zv
    @@ -367,4 +373,31 @@ pub fn set_buffer(buffer_or_name: LispObject) -> LispObject {
         buffer
     }
     
    +/// Signal a `buffer-read-only' error if the current buffer is read-only.
    +/// If the text under POSITION (which defaults to point) has the
    +/// `inhibit-read-only' text property set, the error will not be raised.
    +#[lisp_fn(min = "0")]
    +pub fn barf_if_buffer_read_only(position: Option<EmacsInt>) -> () {
    +    let pos = match position {
    +        None => point(),
    +        Some(n) => {
    +            if n < 0 {
    +                wrong_type!(Qintegerp, LispObject::from_fixnum(n));
    +            } else {
    +                n
    +            }
    +        }
    +    };
    +
    +    let buffer = ThreadState::current_buffer();
    +    let inhibit_read_only = LispObject::from_raw(unsafe { globals.f_Vinhibit_read_only });
    +    let prop = LispObject::from_raw(unsafe {
    +        Fget_text_property(LispObject::from(pos).to_raw(), Qinhibit_read_only, Qnil)
    +    });
    +
    +    if buffer.is_read_only() && inhibit_read_only.is_nil() && prop.is_nil() {
    +        xsignal!(Qbuffer_read_only, current_buffer())
    +    }
    +}
    +
     include!(concat!(env!("OUT_DIR"), "/buffers_exports.rs"));
    diff --git a/src/buffer.c b/src/buffer.c
    index c37ba33..041807e 100644
    --- a/src/buffer.c
    +++ b/src/buffer.c
    @@ -2045,25 +2045,6 @@ set_buffer_if_live (Lisp_Object buffer)
         set_buffer_internal (XBUFFER (buffer));
     }
     
    -DEFUN ("barf-if-buffer-read-only", Fbarf_if_buffer_read_only,
    -				   Sbarf_if_buffer_read_only, 0, 1, 0,
    -       doc: /* Signal a `buffer-read-only' error if the current buffer is read-only.
    -If the text under POSITION (which defaults to point) has the
    -`inhibit-read-only' text property set, the error will not be raised.  */)
    -  (Lisp_Object position)
    -{
    -  if (NILP (position))
    -    XSETFASTINT (position, PT);
    -  else
    -    CHECK_NUMBER (position);
    -
    -  if (!NILP (BVAR (current_buffer, read_only))
    -      && NILP (Vinhibit_read_only)
    -      && NILP (Fget_text_property (position, Qinhibit_read_only, Qnil)))
    -    xsignal1 (Qbuffer_read_only, Fcurrent_buffer ());
    -  return Qnil;
    -}
    -
     DEFUN ("erase-buffer", Ferase_buffer, Serase_buffer, 0, 0, "*",
            doc: /* Delete the entire contents of the current buffer.
     Any narrowing restriction in effect (see `narrow-to-region') is removed,
    @@ -6097,7 +6078,6 @@ Functions running this hook are, `get-buffer-create',
       defsubr (&Skill_buffer);
       defsubr (&Sbury_buffer_internal);
       defsubr (&Sset_buffer_major_mode);
    -  defsubr (&Sbarf_if_buffer_read_only);
       defsubr (&Serase_buffer);
       defsubr (&Sbuffer_swap_text);
       defsubr (&Sset_buffer_multibyte);

**Pull request merged**